lib/streamlit/runtime/session_manager.py

```diff
@@ -0,0 +1,338 @@
+# Copyright (c) Streamlit Inc. (2018-2022) Snowflake Inc. (2022)
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from abc import abstractmethod
+from dataclasses import dataclass
+from typing import Dict, List, Optional
+
+from typing_extensions import Protocol
+
+from streamlit.proto.ForwardMsg_pb2 import ForwardMsg
+from .app_session import AppSession
+
+
+class SessionClientDisconnectedError(Exception):
+    """Raised by operations on a disconnected SessionClient."""
+
+
+class SessionClient(Protocol):
+    """Interface for sending data to a session's client."""
+
+    @abstractmethod
+    def write_forward_msg(self, msg: ForwardMsg) -> None:
+        """Deliver a ForwardMsg to the client.
+
+        If the SessionClient has been disconnected, it should raise a
+        SessionClientDisconnectedError.
+        """
+        raise NotImplementedError
+
+
+# TODO(vdonato): We may want to consider either coming up with a better name for this or
+# consolidating it with the SessionData class since the two names are similar enough
+# to be confusing right now.
+@dataclass
+class SessionInfo:
+    """Type containing data related to an AppSession.
+
+    For each AppSession, the Runtime tracks that session's
+    script_run_count. This is used to track the age of messages in
+    the ForwardMsgCache.
+    """
+
+    client: Optional[SessionClient]
```

Comment:
Making this `Optional` is the part of this PR I'm most iffy about, but I think this won't be an issue in practice until we have more than one `SessionManager` implementation, at which point we'll have a better idea of what our use-cases will be and can tweak these abstractions accordingly (since they won't be hard to change until we have >1 `SessionManager`).

The issue is that I want to have all of the `SessionStore` methods operate on the `SessionInfo` type so that their signatures match that of `SessionManager`, but "inactive" sessions in a `SessionStore` may not have a `SessionClient` attached to them.

We could get around this problem by having some sort of `InactiveSessionInfo` type, but the difficulty with that solution is that we lose the nice behavior where `SessionManager`s that don't differentiate between active and inactive sessions are able to rely on default implementations to automatically define a bunch of methods for them.

Reply:
Ack --

I'll defer changing this for now because I have some ideas of how to make this better at the type level. The main idea would be to enforce that anything that's returned by an active-session version of a method must have a well-defined `client` field, but the non-active session versions of the methods won't require this.

I'm not 100% sure this is sane, though, so I'll try it in a separate PR to the feature branch.